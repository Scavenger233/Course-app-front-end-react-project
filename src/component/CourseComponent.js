import React, { Component } from 'react';
import withRouter from './withRouter';
//To use formik to edit form of course details, this is the most populare framework with React.
import { Formik, Form, Field, ErrorMessage } from 'formik';
import CourseDataService from '../service/CourseDataService';
const INSTRUCTOR = 'bytecaptain'

class CourseComponent extends Component {
    constructor(props) {
        super(props)
        this.state = {
            id: this.props.match.params.id,
            description: ''
        }
        this.onSubmit = this.onSubmit.bind(this)
        this.validate = this.validate.bind(this)
    }

    componentDidMount() {
        console.log(this.state.id)
        // eslint-disable-next-line
        debugger
        if (this.state.id === -1) {
            return
        }
        CourseDataService.retrieveCourse(INSTRUCTOR, this.state.id)
            .then(response => this.setState({
                description: response.data.description
            })).catch(error => {
                //TODO better handle errors
                return error;
            })

    }

    //Checking for empty description
    //Checking for a minimum length of 5
    validate(values) {
        let errors = {}
        if (!values.description) {
            errors.description = 'Enter a Description'
        } else if (values.description.length < 5) {
            errors.description = 'Enter at least 5 Characters in Description'
        }

        return errors
    }

    onSubmit(values) {
        let username = INSTRUCTOR

        let course = {
            id: this.state.id,
            description: values.description,
            targetDate: values.targetDate
        }

        if (this.state.id === -1) {
            CourseDataService.createCourse(username, course)
                .then(() => this.props.navigation('/courses'))
                .catch(error => {
                    //TODO better handle errors https://axios-http.com/docs/handling_errors
                    return error;
                })
        } else {
            CourseDataService.updateCourse(username, this.state.id, course)
                .then(() => this.props.navigation('/courses'))
                .catch(error => {
                    //TODO better handle errors https://axios-http.com/docs/handling_errors
                    return error;
                })
        }

        console.log(values);
    }

    render() {
        //Creating local variable using destructing
        let { description, id } = this.state
        return (
            <div>
                <h3>Course</h3>
                <div data-testid="courseContainer" className="container">
                    {/* Initialing Formik with the values loaded from state */}
                    <Formik
                        initialValues={{ id, description }}
                        onSubmit={this.onSubmit}
                        validateOnChange={true}
                        validateOnBlur={true}
                        validate={this.validate}
                        enableReinitialize={true}>
                        {
                            (props) => (
                                <Form>
                                    {/* Error message alert during description */}
                                    <ErrorMessage name="description" component="div"
                                        className="alert alert-warning" />
                                    <fieldset className="form-group">
                                        <label>Id</label>
                                        {/* Creating a disabled text element for id. The name of element should match the name in state. */}
                                        <Field data-testid="courseId" className="form-control" type="text" name="id" disabled />
                                    </fieldset>
                                    <fieldset className="form-group">
                                        <label>Description</label>
                                        <Field data-testid="courseDescription" className="form-control" type="text" name="description" />
                                    </fieldset>
                                    {/* Adding a submit buttom. */}
                                    <button className="btn btn-success" type="submit">Save</button>
                                </Form>
                            )
                        }
                    </Formik>

                </div>
            </div>
        )
    }
}

export default withRouter(CourseComponent);