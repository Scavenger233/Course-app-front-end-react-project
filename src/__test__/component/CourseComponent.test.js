import React from 'react';

import axios from 'axios';

import CourseComponent from '../.././component/CourseComponent';
import renderer from 'react-test-renderer';

import { BrowserRouter as Router} from 'react-router-dom';

import { fireEvent, render, screen, waitFor } from '@testing-library/react';

jest.mock('axios');

test("renders correctly", async () => {

    const courses = [{ "id": 10001, "username": "in28minutes", "description": "Learn JPA" }, { "id": 10002, "username": "in28minutes", "description": "Learn JPA Course" }];
    axios.get.mockResolvedValue(courses);

    const addParams = {params: {id: -1}};
    const addCourseComponent = renderer.create(<Router><CourseComponent match={addParams}/></Router>).toJSON();
    expect(addCourseComponent).toMatchSnapshot();

});

test('add course', async () => {

    const params = {params: {id: -1}};
    render(<Router><CourseComponent match={params} /></Router>);
    
    const descriptionField = screen.getByTestId('courseDescription');
    await waitFor(() => descriptionField);
    fireEvent.blur(descriptionField);

    const courseContainer = screen.getByTestId('courseContainer');
    await waitFor(() => {
        expect(courseContainer).toHaveTextContent("Enter a Description");  
    });

    fireEvent.change(descriptionField, { target: { value: '112' }});

    await waitFor(() => {
        expect(courseContainer).toHaveTextContent("Enter at least 5 Characters in Description");  
    }); 

    fireEvent.change(descriptionField, { target: { value: 'Description' }});
    
    await waitFor(() => {
        const descriptionError = screen.queryByText("Enter at least 5 Characters in Description");
        expect(descriptionError).toBeNull();
    }); 
    
    let course = { "id": 10001, "username": "bytecaptain", "description": "Learn JPA" };
    axios.post.mockResolvedValue(course);

    const button = screen.getByText("Save");

    await waitFor(() => {
        fireEvent.click(button);
    });  
    
});

test('update course', async () => {

    //TODO refactor to be reused
    const updatedcourse = {data: { "id": 10001, "username": "bytecaptain", "description": "Learn JPA" }};
    axios.get.mockResolvedValue(updatedcourse);

    const params = {params: {id: 10001}};
    render(<Router><CourseComponent match={params} /></Router>);
    
    const descriptionField = screen.getByTestId('courseDescription');
    await waitFor(() => descriptionField);

    fireEvent.change(descriptionField, { target: { value: 'Learn JPA Description updated' }});
    
    let course = { "id": 10001, "username": "bytecaptain", "description": "Learn JPA Description updated" };
    axios.put.mockResolvedValue(course);

    const button = screen.getByText("Save");

    await waitFor(() => {
        fireEvent.click(button);
    });    
    
});
