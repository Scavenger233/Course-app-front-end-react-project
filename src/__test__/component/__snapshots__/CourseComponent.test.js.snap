// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <h3>
    Course
  </h3>
  <div
    className="container"
    data-testid="courseContainer"
  >
    <form
      action="#"
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <fieldset
        className="form-group"
      >
        <label>
          Id
        </label>
        <input
          className="form-control"
          data-testid="courseId"
          disabled={true}
          name="id"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value={-1}
        />
      </fieldset>
      <fieldset
        className="form-group"
      >
        <label>
          Description
        </label>
        <input
          className="form-control"
          data-testid="courseDescription"
          name="description"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value=""
        />
      </fieldset>
      <button
        className="btn btn-success"
        type="submit"
      >
        Save
      </button>
    </form>
  </div>
</div>
`;
